### ISSUE TITLE
New endpoint for user administration (CRUD)

### ISSUE THREAD

**Issue by AdriiiPRodri** [wazuh/wazuh#3280]
## Introduction
Hi team,
In API 3.9 there is a script [configure_api.sh](https://github.com/wazuh/wazuh-api/blob/master/scripts/configure_api.sh) that allows you to configure a single user, the purpose of this issue is to create the endpoints for the **CRUD** cycle of users.
## Restrictions
- The two users created by default (wazuh, wazuh-api) should not be allowed to be deleted.
- Changes must be propagated between all nodes in cluster, all nodes must have a copy of users information.
## Design
In order to propagate the changes to all the nodes of the cluster two options are considered:
- A distributed call that executes the same sentence in all the nodes of the cluster, the main disadvantage of this solution is that if a node is down or has no connection at the time when users are modified. This will be without copy of users, its advantage is that the implementation is simple.
- As a result of the solution commented before, to solve its inconvenience we can create a new task in the cluster that updates the information of the users database like [agent info thread](https://documentation.wazuh.com/3.x/user-manual/manager/wazuh-cluster.html) nodes every certain time and this way each node has the updated information.
- Use [rqlite](https://github.com/rqlite/rqlite)
#### User administration
- Get the information of all users, **the password must not appear**.
    - GET security/users
- Get the information of the specified user, the password must not appear.
    - GET security/users/{user_id}
- Create a new user, the content of the body must be (username: "username", password: "password")
    - POST security/users
- Modify a specified user, the content of the body must be (password: "password")
    - PUT security/users/{user_id}
- Delete a specified user
    - DELETE security/users/{user_id}
### Coding
To enable all these features we will have to create a new file called user_manager.py in the framework/wazuh folder, this file will have the following functions:
- This function will provide as much information as possible about the users:
```python
def get_users(pretty=False, wait_for_complete=False, limit=None, search=None, sort=None)
```
And one example of its output:
```bash
{
  "data": {
    "items": [
      {
        "username": "wazuh"
      },
      {
        "username": "wazuh-api"
      }
    ],
    "totalItems": 2
  }
}
```
- This function will provide as much information as possible about one specified user:
```python
def get_user_id(user_id, pretty=False, wait_for_complete=False)
```
And one example of its output:
```bash
{
  "data": {
    "items": [
      {
        "username": "wazuh"
      }
    ],
    "totalItems": 2
  }
}
```
- This function will create a user with the data that we provide in the body of the request.
```python
def create_user(username, password, pretty=False, wait_for_complete=False)
```
The body of the request must be: 

> "username": "new", "password": "newpass"

And one example of its output:
```bash
{
  "data": {
    "message": "User created correctly"
  }
}
```
- This function will update an specified user with the data that we provide in the body of the request.
```python
def update_user(password, pretty=False, wait_for_complete=False)
```
The body of the request must be:  
> "password": "newpassupdated"

And one example of its output:
```bash
{
  "data": {
    "message": "User updated correctly"
  }
}
```
- This function will delete an specified user.
```python
def delete_user(username, pretty=False, wait_for_complete=False)
```
And one example of its output:
```bash
{
  "data": {
    "message": "User deleted correctly"
  }
}
```
### Script configure_api.sh
As for the functionality available in [configure_api.sh](https://github.com/wazuh/wazuh-api/blob/master/scripts/configure_api.sh) to change port, add a proxy to the api and enable or disable https, it will be updated so that the changes are applied in the new API. Changes should be made to uWSGI so update the script for the new components.

We have to update the paths and adapt the rules to the new format included in [uwsig.yaml](https://github.com/wazuh/wazuh/blob/dev-flask-poc/api/api/configuration/uwsgi.yaml)

### ToDo

- [x] Implement CRUD endpoints

- [x] Update the configure_api.sh file

Best regards,
Adri

**Comment by AdriiiPRodri** [comment]
## Status update
The endpoint has been implemented for the users creation, with a distributed call:
```{bash}
root@243c717a535e:/# curl -u wazuh:wazuh -X POST -H "Content-Type":"application/json" -d '{"username":"new","password":"newpass"}' "http://localhost:55000/security/new_user?username=aaad&password=a"
{
  "data": {
    "message": "User created correctly"
  }
}
```
It remains to be determined what solution we are going to implement.

**Comment by AdriiiPRodri** [comment]
# Status update
## Summary
This development is under continuous changes due to its dependence. 

The necessary endpoints for the users CRUD cycle have been implemented, in a beginning the endpoints were coded as distributed calls through the cluster.

The algorithm used to generate the hashes of the users passwords is [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2) which by its own nature generated different hashes in all nodes.

Finally, since the workers requests are redirected to the master, we have decided to change the calls to a local_master model and this will be the only one that will have the users file.

When we switch to a distributed master model this will also be changed, it is prepared for this purpose.

## Tasks
* [2.5h] Investigate about the wazuh cluster, the password encryption used and the possibility of using rqlite. 
* [2.5h] Finish the CRUD cycle entrypoints

## Difficulties
* Understand the cluster

## Tests performed
These are the tests performed on the CRUD cycle:
#### Get all users:
```bash
root@8aa047d79ba1:/# curl -u wazuh:wazuh -X GET "http://localhost:55000/security/users"
{
  "data": {
    "items": [
      {
        "username": "wazuh-app"
      },
      {
        "username": "wazuh"
      }
    ],
    "totalItems": 2
  }
}
```
#### Get a specified user:
```bash
root@8aa047d79ba1:/# curl -u wazuh:wazuh -X GET "http://localhost:55000/security/users/wazuh"
{
  "data": {
    "items": [
      {
        "username": "wazuh"
      }
    ],
    "totalItems": 1
  }
}
```
#### Create an user called newuser:
```bash
root@8aa047d79ba1:/# curl -u wazuh:wazuh -X POST -H "Content-Type":"application/json" -d '{"username":"newuser","password":"newuser"}' "http://localhost:55000/security/users"
{
  "data": {
    "message": "User created correctly"
  }
}
```
#### Database after the creation of newuser:
```bash
root@8aa047d79ba1:/# sqlite3 /var/ossec/api/configuration/security/users.db SQLite version 3.22.0 2018-01-22 18:45:57
Enter ".help" for usage hints.
sqlite> select * from users;
wazuh-app|pbkdf2:sha256:150000$UODmBzYW$a8b32fda399339c284189b2062d37ffa7fb42ba0c2022ef69e4a05cb17bca03a
wazuh|pbkdf2:sha256:150000$VSnIDKBr$3c85a165bb78d0188d9bbea74379f75653231bdfecfd6bfdcf6dd2629301e8c6
newuser|pbkdf2:sha256:150000$wAKTR31h$78fd40a560a72b3bca773f2048a9b6a8f404ca8a9edf84d00de660a829132f3d
sqlite> .quit
```
#### Get all users after newuser creation:
```bash
root@8aa047d79ba1:/# curl -u wazuh:wazuh -X GET "http://localhost:55000/security/users"
{
  "data": {
    "items": [
      {
        "username": "wazuh-app"
      },
      {
        "username": "wazuh"
      },
      {
        "username": "newuser"
      }
    ],
    "totalItems": 3
  }
}
```
#### Modify newuser:
```bash
root@8aa047d79ba1:/# curl -u wazuh:wazuh -X PUT -H "Content-Type":"application/json" -d '{"password":"newusermodified"}' "http://localhost:55000/security/users/newuser"{
  "data": {
    "message": "User updated correctly"
  }
}
```
#### Database after the modification:
```bash
root@8aa047d79ba1:/# sqlite3 /var/ossec/api/configuration/security/users.db SQLite version 3.22.0 2018-01-22 18:45:57
Enter ".help" for usage hints.
sqlite> select * from users;
wazuh-app|pbkdf2:sha256:150000$UODmBzYW$a8b32fda399339c284189b2062d37ffa7fb42ba0c2022ef69e4a05cb17bca03a
wazuh|pbkdf2:sha256:150000$VSnIDKBr$3c85a165bb78d0188d9bbea74379f75653231bdfecfd6bfdcf6dd2629301e8c6
newuser|pbkdf2:sha256:150000$vjTCclWn$2ba87f54415fb9d50b221915ddbad454ef9ad931721fe4570469719e124cb89f
sqlite> .quit
```
#### Delete new user:
```bash
root@8aa047d79ba1:/# curl -u wazuh:wazuh -X DELETE "http://localhost:55000/security/users/newuser"
{
  "data": {
    "message": "User deleted correctly"
  }
}
```
#### Database after the delete:
```bash
root@8aa047d79ba1:/# sqlite3 /var/ossec/api/configuration/security/users.db SQLite version 3.22.0 2018-01-22 18:45:57
Enter ".help" for usage hints.
sqlite> select * from users;
wazuh-app|pbkdf2:sha256:150000$UODmBzYW$a8b32fda399339c284189b2062d37ffa7fb42ba0c2022ef69e4a05cb17bca03a
wazuh|pbkdf2:sha256:150000$VSnIDKBr$3c85a165bb78d0188d9bbea74379f75653231bdfecfd6bfdcf6dd2629301e8c6
```
#### Trying to delete wazuh user
```bash
root@8aa047d79ba1:/# curl -u wazuh:wazuh -X DELETE "http://localhost:55000/security/users/wazuh"
{
  "data": {
    "message": "The users wazuh and wazuh-app can not be deleted"
  }
}
root@8aa047d79ba1:/# sqlite3 /var/ossec/api/configuration/security/users.db SQLite version 3.22.0 2018-01-22 18:45:57
Enter ".help" for usage hints.
sqlite> select * from users;
wazuh-app|pbkdf2:sha256:150000$UODmBzYW$a8b32fda399339c284189b2062d37ffa7fb42ba0c2022ef69e4a05cb17bca03a
wazuh|pbkdf2:sha256:150000$VSnIDKBr$3c85a165bb78d0188d9bbea74379f75653231bdfecfd6bfdcf6dd2629301e8c6
```

## Pending
* Improve endpoint definition

**Comment by AdriiiPRodri** [comment]
## Status update
I have implemented a new script in Python that improves the functionality of the old configure_api.sh, now the script is launched with parameters and not by questions which has greater security risk.

The functionality of configuring https with the script is not available, this is blocked by #3054  

Once uWSGI allows this, this function will be added to the script.

## Stoppers
- uWSGI issue #3054 

## Demostration
```bash
root@572894e1bd3b:/# /var/ossec/framework/python/bin/python3 /var/ossec/api/scripts/configure_api.py -h                                           
usage: configure_api.py [-h] [-p PORT] [-i IP] [-s HTTPS] [-b BASIC]
                        [-bp PROXY]

optional arguments:
  -h, --help            show this help message and exit
  -p PORT, --port PORT  Change port number
  -i IP, --ip IP        Change the host IP
  -s HTTPS, --https HTTPS
                        Enable https protocol (TRUE/FALSE)
  -b BASIC, --basic BASIC
                        Configure basic authentication
  -bp PROXY, --proxy PROXY
                        Yes to run API behind a proxy
```
```bash
root@572894e1bd3b:/# /var/ossec/framework/python/bin/python3 /var/ossec/api/scripts/configure_api.py -i 192.168.0.1 -p 20 -s yeS -b true -bp false
[INFO] IP changed correctly
[INFO] PORT changed correctly
[INFO] PROXY value changed correctly
[INFO] Basic auth value changed correctly
```
```bash
root@572894e1bd3b:/# cat /var/ossec/api/configuration/uwsgi.yaml       
# uWSGI server configuration
uwsgi:

  # HTTP (comment the following line to disable non secure http connections)
  http:  192.168.0.1:20
  # HTTPS (uncoment the following lines and get the needed certificates in order to enable https)
  # More details in https://uwsgi-docs.readthedocs.io/en/latest/HTTPS.html
  # shared-socket: 0.0.0.0:443
  # https: =0,foobar.crt,foobar.key,HIGH #,ca.crt

  # Set parallel processes and threads for each process (1 by default)
  # processes: 4
  # threads: 4

  # Set the user running the API server
  uid: ossec
  gid: ossec

  # Default configuration (DO NOT MODIFY)
  single-interpreter: true
  enable-threads: true
  master: 1
  die-on-term: true
```
```bash
root@572894e1bd3b:/# cat /var/ossec/api/configuration/api.yaml         
# USE THIS FILE AS A TEMPLATE. COPY IT TO A NEW api.yml FILE IN THIS FOLDER AND MODIFY AT WILL

# Use HTTP authentication
basic_auth: yes
# Set this option to "yes" in case the API is running behind a proxy server. Values: yes, no
behind_proxy_server: no

# Advanced configuration

# Logging configuration
# logs:
  # Values for API log level: disabled, info, warning, error, debug, debug2 (each level includes the previous level).
  # level: "info"
  # log path
  # path: "logs/api.log"

# Cross-origin resource sharing. Values: yes, no.
# cors: yes

# Cache (time in milliseconds)
# cache:
  # enabled: yes
  # debug: no
  # time: 750

# Force the use of authd when adding and removing agents. Values: yes, no
# use_only_authd: no

# Drop privileges (Run as ossec user)
# drop_privileges: yes

# Enable features under development
# experimental_features: no
```

**Comment by AdriiiPRodri** [comment]
## Status update
The script has been improved, now in besides being administrated by parameters, we can also use a easier interactive mode which will ask questions until the API is fully configured.

Currently this issue still has stoppers, at the moment we can not configure the user by enabling basic auth. The final design of the configuration files may change and therefore the script is still susceptible to changes.

New parameters have been added to the script:

```bash
adriiiprodri@Wazuh:/# /var/ossec/framework/python/bin/python3 /var/ossec/api/scripts/configure_api.py -h 
usage: configure_api.py [-h] [-p PORT] [-i IP] [-b BASIC] [-x PROXY] [-t HTTP]
                        [-s HTTPS] [-I]

optional arguments:
  -h, --help            show this help message and exit
  -p PORT, --port PORT  Change port number
  -i IP, --ip IP        Change the host IP
  -b BASIC, --basic BASIC
                        Configure basic authentication (true/false)
  -x PROXY, --proxy PROXY
                        Yes to run API behind a proxy
  -t HTTP, --http HTTP  Enable http protocol (true/false)
  -s HTTPS, --https HTTPS
                        Enable https protocol (true/false)
  -I, --interactive     Enables guided configuration
```

The messages have been improved, now it is not possible to write a wrong value in the configuration files.

## Stoppers

- uWSGI issue #3054

**Comment by AdriiiPRodri** [comment]
## Status update
Several exceptions have been added to control unexpected situations. Error messages have also been improved, which have become **WazuhError** or **WazuhInternalError**.

The class for user administration has been improved, **reducing** lines of code and improving performance. In addition the messages that are now obtained when users are got, created, modified or deleted have been improved.

Added the first tests for the **GETs** of the CRUD user cycle.

### Integrated tests
```bash
adriiiprodri@Wazuh:/# /var/ossec/framework/python/bin/tavern-ci -vv /var/ossec/framework/python/bin/test/test_users_GET_endpoints.tavern.yaml 
======================================================================= test session starts ========================================================================
platform linux -- Python 3.7.2, pytest-4.6.4, py-1.8.0, pluggy-0.12.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /var/ossec/framework/python/bin/test
plugins: tavern-0.26.4
collected 2 items                                                                                                                                                  

var/ossec/framework/python/bin/test/test_users_GET_endpoints.tavern.yaml::GET /security/users PASSED                                                         [ 50%]
var/ossec/framework/python/bin/test/test_users_GET_endpoints.tavern.yaml::GET /security/users/{username} PASSED                                              [100%]

========================================================================= warnings summary =========================================================================
var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126
  /var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    if not isinstance(key, collections.Hashable):

test_users_GET_endpoints.tavern.yaml::GET /security/users
test_users_GET_endpoints.tavern.yaml::GET /security/users/{username}
  /var/ossec/framework/python/lib/python3.7/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=============================================================== 2 passed, 3 warnings in 0.58 seconds ===============================================================
```

**Comment by AdriiiPRodri** [comment]
## Status update

Several functions and parts of the code that were unreachable have been removed, missing docstring have been added.

Several uncontrolled situations have been detected, in order to solve this, several **exceptions** have been added to control the situation.

A new requirement has been added for user passwords, **the minimum length they must have is 8 characters and must contain at least one capital letter, one number and one special character**. With this requirement we will increase the security of the API. The SPEC has been modified to reflect these changes.

Finally, a **new system of privileges** has been added within the users themselves. These layers are divided into users with **administrator capabilities and normal or non-administrator users**. The difference between the two is that administrator users will be able to modify other users, even turn them into administrators, however, non-administrator users will only be able to modify themselves and in any case another user.

**Comment by AdriiiPRodri** [comment]
## Status update

Tests for **POST, PUT and DELETE** methods have been added. 

In the process, two bugs have been fixed when controlling exceptions. These caused an exception to be raised which was not really the one it should have been.

## Tests results
### POST
```bash
adriiiprodri@Wazuh python3 -m pytest -vv test_users_POST_endpoints.tavern.yaml  
====================================================================================== test session starts =======================================================================================
platform linux -- Python 3.5.3, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/adriiiprodri/Desktop/git/wazuh/api
plugins: tavern-0.28.0
collected 1 item                                                                                                                                                                                 

test_users_POST_endpoints.tavern.yaml::POST /security/users PASSED                                                                                                                         [100%]

============================================================================== 1 passed, 1 warnings in 0.94 seconds ==============================================================================
```

### PUT
```bash
adriiiprodri@Wazuh python3 -m pytest -vv test_users_PUT_endpoints.tavern.yaml 
====================================================================================== test session starts =======================================================================================
platform linux -- Python 3.5.3, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/adriiiprodri/Desktop/git/wazuh/api
plugins: tavern-0.28.0
collected 1 item                                                                                                                                                                                 

test_users_PUT_endpoints.tavern.yaml::PUT /security/users/{username} PASSED                                                                                                                          [100%]

============================================================================== 1 passed, 1 warnings in 0.77 seconds ==============================================================================
```

### DELETE
```bash
adriiiprodri@Wazuh python3 -m pytest -vv test_users_DELETE_endpoints.tavern.yaml
====================================================================================== test session starts =======================================================================================
platform linux -- Python 3.5.3, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/adriiiprodri/Desktop/git/wazuh/api
plugins: tavern-0.28.0
collected 1 item                                                                                                                                                                                 

test_users_DELETE_endpoints.tavern.yaml::DELETE /security/users PASSED                                                                                                                     [100%]

============================================================================== 1 passed, 1 warnings in 0.54 seconds ==============================================================================
```
